constraint/pkg/client/crd_helpers.go

```diff
@@ -60,49 +71,62 @@ func newCRDHelper() *crdHelper {
 
 // createCRD takes a template and a schema and converts it to a CRD
 func (h *crdHelper) createCRD(
-	templ *v1alpha1.ConstraintTemplate,
-	schema *apiextensionsv1beta1.JSONSchemaProps) *apiextensionsv1beta1.CustomResourceDefinition {
-	crd := &apiextensionsv1beta1.CustomResourceDefinition{
-		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+	templ *templates.ConstraintTemplate,
+	schema *apiextensions.JSONSchemaProps) (*apiextensions.CustomResourceDefinition, error) {
+	crd := &apiextensions.CustomResourceDefinition{
+		Spec: apiextensions.CustomResourceDefinitionSpec{
 			Group: constraintGroup,
-			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+			Names: apiextensions.CustomResourceDefinitionNames{
 				Kind:     templ.Spec.CRD.Spec.Names.Kind,
 				ListKind: templ.Spec.CRD.Spec.Names.Kind + "List",
 				Plural:   strings.ToLower(templ.Spec.CRD.Spec.Names.Kind),
 				Singular: strings.ToLower(templ.Spec.CRD.Spec.Names.Kind),
 			},
-			Validation: &apiextensionsv1beta1.CustomResourceValidation{
+			Validation: &apiextensions.CustomResourceValidation{
 				OpenAPIV3Schema: schema,
 			},
 			Scope:   "Cluster",
-			Version: v1alpha1.SchemeGroupVersion.Version,
+			Version: v1beta1.SchemeGroupVersion.Version,
+			Versions: []apiextensions.CustomResourceDefinitionVersion{
+				{
+					Name:    v1beta1.SchemeGroupVersion.Version,
+					Storage: true,
+					Served:  true,
+				},
+				{
+					Name:    v1alpha1.SchemeGroupVersion.Version,
+					Storage: false,
+					Served:  true,
+				},
+			},
 		},
 	}
-	h.scheme.Default(crd)
-	crd.ObjectMeta.Name = fmt.Sprintf("%s.%s", crd.Spec.Names.Plural, constraintGroup)
-	return crd
+	// Defaulting functions only exist for v1beta1
```

Comment:
can you provide more information for this?

Reply:
The kubernetes scheme is documented here:

https://godoc.org/k8s.io/apimachinery/pkg/runtime#Scheme

I've tried to use the unversioned, internal representation of the CRD types in an effort to make the API a bit more stable.

The CRD library has chosen to only register defaulting functions (those that set defaults on a type) for the versioned verision of the CRD resource. This function takes an unversioned schema, which corresponds to the unversioned CRD.

This means I have two options:

1. Convert the schema to a versioned JSONSchema, implement a versioned CRD, apply defaults, convert to unversioned.
2. Create an unversioned CRD, convert to versioned CRD, apply defaults, convert to unversioned CRD

I expect choice (2) will require less change if/when CRD v2 comes around.